modules/mqtt/destination/mqtt-worker.c

```diff
@@ -0,0 +1,275 @@
+      return LTR_NOT_CONNECTED;
+
+    case MQTTCLIENT_MAX_MESSAGES_INFLIGHT:
+      msg_error("Max message inflight! (publish)",
+                log_pipe_location_tag(&self->owner->super.super.super));
+      return LTR_ERROR;
+
+    case MQTTCLIENT_FAILURE:
+      msg_error("Failure during publishing!",
+                log_pipe_location_tag(&self->owner->super.super.super));
+      return LTR_ERROR;
+
+    case MQTTCLIENT_PERSISTENCE_ERROR:
+      g_assert_not_reached();
+    case MQTTCLIENT_BAD_QOS:
+      g_assert_not_reached();
+    case MQTTCLIENT_BAD_STRUCTURE:
+      g_assert_not_reached();
+    default:
+      g_assert_not_reached();
+
+    case MQTTCLIENT_NULL_PARAMETER:
+    case MQTTCLIENT_BAD_UTF8_STRING:
+      msg_error("Error during publish!",
+                evt_tag_str("error code", MQTTClient_strerror(result)),
+                log_pipe_location_tag(&self->owner->super.super.super));
+      return LTR_DROP;
+    }
+}
+
+static LogThreadedResult
+_wait_result_evaluation(LogThreadedDestWorker *self, gint result)
+{
+  switch(result)
+    {
+    case MQTTCLIENT_SUCCESS:
+      return LTR_SUCCESS;
+
+    case MQTTCLIENT_DISCONNECTED:
+      msg_error("Disconnected while waiting the response!",
+                log_pipe_location_tag(&self->owner->super.super.super));
+      return LTR_NOT_CONNECTED;
+
+    case MQTTCLIENT_FAILURE:
+    default:
+      msg_error("Error while waiting the response!",
+                evt_tag_str("error code", MQTTClient_strerror(result)),
+                log_pipe_location_tag(&self->owner->super.super.super));
+      return LTR_ERROR;
+    }
+}
+
+static const gchar *
+mqtt_dest_worker_resolve_template_topic_name(MQTTDestinationWorker *self, LogMessage *msg)
+{
+  MQTTDestinationDriver *owner = (MQTTDestinationDriver *) self->super.owner;
+
+  LogTemplateEvalOptions options = {&owner->template_options, LTZ_SEND, self->super.seq_num, NULL};
+  log_template_format(owner->topic_name, msg, &options, self->topic_name_buffer);
+
+  GError *error = NULL;
+
+  if (mqtt_dd_validate_topic_name(self->topic_name_buffer->str, &error))
+    {
+      return self->topic_name_buffer->str;
+    }
+
+  msg_error("Error constructing topic", evt_tag_str("topic_name", self->topic_name_buffer->str),
+            evt_tag_str("driver", owner->super.super.super.id),
+            log_pipe_location_tag(&owner->super.super.super.super),
+            evt_tag_str("error message", error->message));
+
+  g_error_free(error);
+
+  return owner->fallback_topic_name;
+}
+
+static LogThreadedResult
+_mqtt_send(MQTTDestinationWorker *self, char *msg, const gchar *topic)
+{
+  MQTTClient_message pubmsg = MQTTClient_message_initializer;
+  MQTTClient_deliveryToken token;
+  gint rc;
+  LogThreadedResult result = LTR_SUCCESS;
+
+  pubmsg.payload = msg;
+  pubmsg.payloadlen = (int)strlen(msg);
+  pubmsg.qos = self->qos;
+  pubmsg.retained = 0;
+
+  rc = MQTTClient_publishMessage(self->client, topic, &pubmsg, &token);
```

Comment:
I think it would be useful to add a debug message about the outgoing message, for example:

```
msg_debug("Outgoing message to MQTT destination", evt_tag_str("topic", topic), evt_tag_str("message", msg)
          log_pipe_location_tag(&self->owner->super.super.super));
```